Move error reporting code out-of-line from header.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32168 Eliminate exception throwing code from dispatch call sites
* #32121 Out-of-line construction of OperatorName.
* **#32118 Move error reporting code out-of-line from header.**
* #32117 Make an assert on a hotpath trigger only in DEBUG mode.

This reduces code size and makes the calling function more likely to inline.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392950](https://our.internmc.facebook.com/intern/diff/D19392950)

ezyang committed:
Move error reporting code out-of-line from header.

This reduces code size and makes the calling function more likely to inline.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move error reporting code out-of-line from header."


This reduces code size and makes the calling function more likely to inline.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit d217f7ef:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d217f7ef4d1763433549c481f9fd154f70bf9320).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4222479) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84353764) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84353764))
<details>
<summary>
<code>Jan 13 19:56:50 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 13 19:56:50 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 13 19:56:50 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 13 19:56:50 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 13 19:56:50 1 out of 1 hunk FAILED 
Jan 13 19:56:50 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 13 19:56:50 Traceback (most recent call last): 
Jan 13 19:56:50   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 13 19:56:50     patch_repo(args) 
Jan 13 19:56:50   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 13 19:56:50     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 13 19:56:50 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 13 19:56:51 + cleanup 
Jan 13 19:56:51 + retcode=1 
Jan 13 19:56:51 + set +x 
Jan 13 19:56:51 =================== sccache compilation log =================== 
Jan 13 19:56:51 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 13 19:56:51 Compile requests                 1 
Jan 13 19:56:51 Compile requests executed        0 
Jan 13 19:56:51 Cache hits                       0 
Jan 13 19:56:51 Cache misses                     0 
Jan 13 19:56:51 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang committed:
Update on "Move error reporting code out-of-line from header."


This reduces code size and makes the calling function more likely to inline.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392950](https://our.internmc.facebook.com/intern/diff/D19392950)

[ghstack-poisoned]

smessmer commented:
good idea

ezyang committed:
Update on "Move error reporting code out-of-line from header."


This reduces code size and makes the calling function more likely to inline.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19392950](https://our.internmc.facebook.com/intern/diff/D19392950)

[ghstack-poisoned]
